qp_shotgun/kneaddata/kneaddata.py

```diff
@@ -0,0 +1,272 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from os.path import basename, join
+
+from qiita_client import ArtifactInfo
+
+from qiita_client.util import system_call, get_sample_names_by_run_prefix
+
+
+def make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file):
+    """Recovers read pairing information
+
+    Parameters
+    ----------
+    forward_seqs : list of str
+        The list of forward seqs filepaths
+    reverse_seqs : list of str
+        The list of reverse seqs filepaths
+    map_file : str
+        The path to the mapping file
+
+    Returns
+    -------
+    samples: list of tup
+        list of 3-tuples with run prefix, sample name, fwd read fp, rev read fp
+
+    Raises
+    ------
+    ValueError
+        If the rev is not an empty list and the same length than fwd seqs
+        The prefixes of the run_prefix don't match the file names
+
+    Notes
+    -----
+    """
+
+    # sort forward seqs
+    forward_seqs.sort()
+
+    # check that rev seqs are same len
+    if reverse_seqs:
+        if len(forward_seqs) != len(reverse_seqs):
+            raise ValueError('Your reverse and forward files are of different '
+                             'length. Forward: %s. Reverse: %s.' %
+                             (', '.join(forward_seqs),
+                              ', '.join(reverse_seqs)))
+        reverse_seqs.sort()
+
+    # get run prefixes
+    # These are prefixes that should match uniquely to forward reads
+    # sn_by_rp is dict of samples keyed by run prefixes
+    sn_by_rp = get_sample_names_by_run_prefix(map_file)
+
+    # make pairings
+    samples = []
+    used_prefixes = []
+    for i, f_fp in enumerate(forward_seqs):
+        # f_fp is the fwd read filepath
+        f_fn = basename(f_fp)
+
+        # iterate over run prefixes and make sure only one matches
+        run_prefix = None
+        for rp in sn_by_rp:
+            if f_fn.startswith(rp) and run_prefix is None:
+                run_prefix = rp
+            elif f_fn.startswith(rp) and run_prefix is not None:
+                raise ValueError('Multiple run prefixes match this fwd read: '
+                                 '%s' % f_fn)
+
+        # make sure that we got one matching run prefix:
+        if run_prefix is None:
+            raise ValueError('No run prefix matching this fwd read: %s'
+                             % f_fn)
+
+        if run_prefix in used_prefixes:
```

Comment:
can `used_prefixes` be a `set`? this will be a list lookup, and while the number of entries probably wont get too large, it's good practice to use a better lookup structure when the thing being searched over is expected to reasonably be > 5 elements